File: server/src/main/java/br/com/diegosilva/automation/actors/Receptor.java
package br.com.diegosilva.automation.actors;

import akka.actor.Cancellable;
import akka.actor.typed.Behavior;
import akka.actor.typed.PreRestart;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;
import akka.cluster.sharding.typed.javadsl.ClusterSharding;
import akka.cluster.sharding.typed.javadsl.EntityRef;
import br.com.diegosilva.automation.CborSerializable;
import br.com.diegosilva.automation.dto.IOTMessage;
import br.com.diegosilva.automation.utils.SerialPortFactory;
import jssc.SerialPort;
import jssc.SerialPortException;

import java.time.Duration;

public class Receptor extends AbstractBehavior<Receptor.Command> {

    private SerialPort serialPort;
    private Cancellable cancellable;

    public static Behavior<Command> create() {
        return Behaviors.setup(context -> new Receptor(context));
    }

    private Receptor(ActorContext<Command> context) throws SerialPortException {
        super(context);
        this.serialPort = SerialPortFactory.get(context.getSystem().settings().config().getString("serial.port"));
    }

    @Override
    public Receive<Command> createReceive() {
        return newReceiveBuilder()
                .onMessage(Start.class, (msg) -> {
                    getContext().getLog().debug("Iniciando escuta");
                    startReceive();
                    return this;
                })
                .onSignal(
                        PreRestart.class,
                        signal -> {
                            getContext().getLog().debug("Reiniciando de recebimento de leituras");
                            this.serialPort = SerialPortFactory.get(getContext().getSystem().settings().config().getString("serial.port"));
                            retry(Duration.ofSeconds(5), new Receptor.Start());
                            return this;
                        })
                .build();
    }

    public void startReceive() {
        try {
            StringBuilder message = new StringBuilder();
            serialPort.addEventListener(event -> {
                if (event.isRXCHAR() && event.getEventValue() > 0) {
                    try {
                        byte buffer[] = serialPort.readBytes();
                        for (byte b : buffer) {
                            if ((b == '\n') && message.length() > 0) {
                                String toProcess = message.toString().replaceAll("[^a-zA-Z0-9:,;._]", "");;
                                getContext().getLog().info("Valor recebido {}", toProcess);
                                message.setLength(0);
                                IOTMessage iotMessage = IOTMessage.decode(toProcess);
                                if (iotMessage != null)
                                    getDevice(iotMessage.id).tell(new Device.Process(iotMessage));
                            } else {
                                message.append((char) b);
                            }
                        }
                    } catch (SerialPortException ex) {
                        System.out.println(ex);
                        System.out.println("serialEvent");
                    }
                }
            });
        } catch (SerialPortException ex) {
            getContext().getLog().error("Erro ao receber leitura", ex);
            throw new RuntimeException(ex);
        }
    }

    private void retry(Duration duration, Command cmd) {
        if (this.cancellable != null) {
            this.cancellable.cancel();
        }
        this.cancellable = getContext().scheduleOnce(duration, getContext().getSelf(), cmd);
    }

    EntityRef<Device.Command> getDevice(String id) {
        ClusterSharding sharding = ClusterSharding.get(getContext().getSystem());
        return sharding.entityRefFor(Device.TypeKey, id);
    }

    public interface Command extends CborSerializable {
    }

    public static class Start implements Command {
    }
}

File: server_scl/src/main/scala/br/com/diegosilva/home/actors/Receptor.scala
package br.com.diegosilva.home.actors

import akka.actor.Cancellable
import akka.actor.typed.scaladsl.{AbstractBehavior, ActorContext, Behaviors}
import akka.actor.typed.{Behavior, PreRestart, Signal}
import akka.cluster.sharding.typed.scaladsl.{ClusterSharding, EntityRef}
import br.com.diegosilva.home.CborSerializable
import br.com.diegosilva.home.actors.Receptor.{Command, Start}
import br.com.diegosilva.home.dto.IOTMessage
import br.com.diegosilva.home.factory.SerialPortFactory
import jssc.SerialPort

import scala.concurrent.duration._

object Receptor {

  sealed trait Command extends CborSerializable

  final case class Start() extends Command

  def apply(): Behavior[Command] =
    Behaviors.setup(context => new Receptor(context))

}

class Receptor(context: ActorContext[Command]) extends AbstractBehavior[Command](context) {

  private var serialPort: SerialPort = SerialPortFactory.get(context.system.settings.config.getString("serial.port"))
  private var cancellable: Cancellable = null

  override def onMessage(msg: Command): Behavior[Command] = {
    case Start =>
      context.log.debug("Iniciando escuta")
      startReceive()
  }

  override def onSignal: PartialFunction[Signal, Behavior[Command]] = {
    case restart: PreRestart => {
      context.log.debug("Reiniciando de recebimento de leituras")
      this.serialPort = SerialPortFactory.get(context.system.settings.config.getString("serial.port"))
      retry(5.seconds, new Receptor.Start)
      Behaviors.same
    }
  }

  private def startReceive() {
    val message = new StringBuilder()
    serialPort.addEventListener { event =>
      if (event.isRXCHAR && event.getEventValue > 0) {
        val buffer: Array[Byte] = serialPort.readBytes()
        buffer.foreach { b =>
          if ((b == '\n') && message.length > 0) {
            val toProcess = message.toString.replaceAll("[^a-zA-Z0-9:,;._]", "")
            context.log.info("Valor recebido {}", toProcess)
            message.setLength(0)
            val iotMessage = IOTMessage.decode(toProcess)
            iotMessage match {
              case Some(iotMessage) =>
                getDevice(iotMessage.id) ! Device.Processar(iotMessage)
            }
          }
          else message.append(b.toChar)
        }
      }
    }
  }

  private def getDevice(id: String): EntityRef[Device.Command] = {
    val sharding = ClusterSharding(context.system)
    sharding.entityRefFor(Device.EntityKey, id)
  }

  private def retry(duration: FiniteDuration, cmd: Receptor.Command): Unit = {
    if (this.cancellable != null) this.cancellable.cancel
    this.cancellable = context.scheduleOnce(duration, context.self, cmd)
  }
}
